src/ixdat/db.py

```diff
@@ -0,0 +1,229 @@
+        self.backend = BackendClass(**db_kwargs)
+
+
+DB = DataBase()  # initate the database. It functions as a global "constant"
+
+
+# THIS is proposed as the main mechanism for changing backend, to make
+# the shared global nature of it explicit. And in any case, the user
+# will never have to deal with the db, except when changing it away
+# from the default. This function should probably be exposed in the
+# top name space.
+
+
+def change_database(db_kind, **db_kwargs):
+    """Change the backend specifying which database objects are saved to/loaded from"""
+    DB.set_backend(db_kind, **db_kwargs)
+
+
+def get_database_kind():
+    """Return the name of the class of which the database backend is an instance"""
+    return DB.backend.__class__.__name__
+
+
+class Saveable:
+    """Base class for table-representing classes implementing database functionality.
+
+    This enables seamless interoperability between databse tables and ixdat classes.
+    Classes inheriting from this need to provide just a bit of info to define the
+    corresponding table, and then then saving and loading should just work.
+
+    At a minimum, the `table_name` and `column_attrs` class attributes need to be
+    overwritten in inheriting classes to define the name and columns of the main
+    corresponding table. If an auxiliary table is needed to store lists of references
+    as rows, this should be represented in `linkers`. Doubly-inheriting classes can use
+    `extra_column_attrs` to add extra columns via an auxiliary table without changing
+    the main table name.
+
+    ixdat is lazy, only loading things when needed. Correspondingly, all of the columns
+    of table mentioned above should refer to (lists of) id's and not actual objects of
+    other ixdat classes.
+
+    The class attributes (defined before __init__) and object attributes (defined in
+    __init__) are described here. See the other methods and the relevant inheriting
+    classes for more info.
+
+    Class attributes:
+        db (DataBase): the database, DB, which has the save, open, and load_data methods
+        table_name (str): The name of the table or folder in which objects are saved
+        column_attrs (dict): {column: attr} where column (str) is the name of the
+            column in the table and attr (str) is the name of the attribute of the
+            class. The two names column and attr are often but not always the same.
+        extra_column_attrs (dict): {table_name: {column: attr}} for auxiliary tables
+            to represent the "extra" attributes, for double-inheriting classes.
+        linkers (dict): {table_name: (reference_table, {column: attr})} for defining
+            the connections between objects.
+
+    Object attributes:
+        backend_name (str): the name of the backend where the object is saved. For a
+            new, un-saved, object, this is "memory".
+        _id (int): the principle key of the object, also accessible as `id`. This should
```

Comment:
Here the function is different. Here it's a managed attribute. `_id` is a private attribute of a `Saveable` class and `id` is the public name of the id...
I've only used `i` as a variable name within a method. This may be an argument against replacing `i` in those cases with `id_` ... it would be confused with `_id` which is used in a different context.

Reply:
I put a TODO in that docstring refering to this thread.